src/sample.cpp

```diff
@@ -12,55 +13,81 @@ const uint64_t constant_rate_hash_factor = 1111111111111111111;
 
 const std::string sample_rate_metric_key = "_sample_rate";
 
-SampleProvider KeepAllSampler() {
-  return {
-      [](SpanContext& context) {
-        context.setBaggageItem(sample_type_baggage_key, "KeepAllSampler");
-        return true;
-      },
-      [](std::unique_ptr<ot::Span>&) {},
-  };
-}
-
-SampleProvider DiscardAllSampler() {
-  return {
-      [](SpanContext& context) {
-        context.setBaggageItem(sample_type_baggage_key, "DiscardAllSampler");
-        return false;
-      },
-      [](std::unique_ptr<ot::Span>&) {},
-  };
-}
-
-SampleProvider ConstantRateSampler(double rate) {
-  uint64_t max_trace_id = 0;
+namespace {
+uint64_t maxIdFromKeepRate(double rate) {
   // This check is required to avoid undefined behaviour converting the rate back from
   // double to uint64_t.
   if (rate == 1.0) {
-    max_trace_id = std::numeric_limits<uint64_t>::max();
+    return std::numeric_limits<uint64_t>::max();
   } else if (rate > 0.0) {
-    max_trace_id = uint64_t(rate * max_trace_id_double);
+    return uint64_t(rate * max_trace_id_double);
+  }
+  return 0;
+}
+}  // namespace
+
+DiscardRateSampler::DiscardRateSampler(double rate)
+    : max_trace_id_(maxIdFromKeepRate(1.0 - rate)) {}
+
+bool DiscardRateSampler::discard(const SpanContext& context) const {
+  // I don't know how voodoo it is to use the trace_id essentially as a source of randomness,
+  // rather than generating a new random number here. It's a bit faster, and more importantly it's
+  // cargo-culted from the agent. However it does still seem too "clever", and makes testing a
+  // bit awkward.
+  uint64_t hashed_id = context.trace_id() * constant_rate_hash_factor;
+  if (hashed_id < max_trace_id_) {
+    return false;
+  }
+  return true;
+}
+
+OptionalSamplingPriority DiscardRateSampler::sample(const std::string& environment,
+                                                    const std::string& service,
+                                                    uint64_t trace_id) const {
+  return nullptr;
+}
+
+bool PrioritySampler::discard(const SpanContext& context) const { return false; }
+
+OptionalSamplingPriority PrioritySampler::sample(const std::string& environment,
+                                                 const std::string& service,
+                                                 uint64_t trace_id) const {
+  uint64_t max_hash;
+  std::ostringstream key;
+  key << "service:" << service << ",env:" << environment;
```

Comment:
Does an empty `environment` need to be rendered as `none` for this key?

Reply:
Nah, that's a Datadog UI convention. Based off of [the PrioritySampling draft RFC](https://docs.google.com/document/d/1wwjpgu7KLZbuVhZAKwY2gGcy4JrvZ9mJo3VSpUOB2AE/edit), just the empty string is used. So a key "service:nginx,env:" is valid (and common!), and leaving env blank (or not even sending it at all) is fine.